asdar/src/objects/int.c

```diff
@@ -68,79 +83,266 @@ static Object *new_from_mpzt(Interp *interp, mpz_t mpz)
 
 Object *intobj_new_long(Interp *interp, long l)
 {
-	mpz_t mpz;
-	mpz_init_set_si(mpz, l);
-	return new_from_mpzt(interp, mpz);
+	long cacheidx;
+	if (l >= 0 && (size_t)l < sizeof(interp->intcache)/sizeof(interp->intcache[0])) {
+		cacheidx = l;
+
+		if (interp->intcache[cacheidx]) {
+			OBJECT_INCREF(interp->intcache[cacheidx]);
+			return interp->intcache[cacheidx];
+		}
+	} else {
+		cacheidx = -1;
+	}
+
+	struct IntData *data = malloc(sizeof *data);
+	if(!data) {
+		errobj_set_nomem(interp);
+		return NULL;
+	}
+
+	data->spilled = false;
+	data->val = l;
+	data->str = NULL;
+
+	Object *res = object_new(interp, &intobj_type, (struct ObjData){
+		.val = data,
+		.destroy = intdata_destroy,
+	});
+	if (res != NULL && cacheidx != -1) {
+		interp->intcache[cacheidx] = res;
+		OBJECT_INCREF(res);
+	}
+	return res;
 }
 
 Object *intobj_new_bebytes(Interp *interp, const unsigned char *seq, size_t len, bool negate)
 {
 	mpz_t mpz;
 	mpz_init(mpz);
-	mpz_import(mpz, len, 1, 1, 0, 0, seq);    // see docs for all the magic numbers
+	mpz_import(
+		mpz,
+		len,
+		1, // order; whole-array endianness
+		1, // size; Size of `seq` element in bytes
+		0, // endian; per-element endianness
+		0, // nails; bits-per-element to skip
+		seq
+	);
+
 	if(negate)
 		mpz_neg(mpz, mpz);
+
 	return new_from_mpzt(interp, mpz);
 }
 
 int intobj_cmp(Object *x, Object *y)
 {
 	assert(x->type == &intobj_type);
 	assert(y->type == &intobj_type);
-	return mpz_cmp( ((struct IntData*)x->data.val)->mpz, ((struct IntData*)y->data.val)->mpz );
+
+	struct IntData *x_data = (struct IntData*) x->data.val;
+	struct IntData *y_data = (struct IntData*) y->data.val;
+
+	if (!x_data->spilled && !y_data->spilled) {
+		/* https://stackoverflow.com/a/10997428 */
+		return (x_data->val > y_data->val) - (x_data->val < y_data->val);
+	} else if (x_data->spilled && !y_data->spilled) {
+		return mpz_cmp_si(x_data->mpz, y_data->val);
+	} else if (!x_data->spilled && y_data->spilled) {
+		return -mpz_cmp_si(y_data->mpz, x_data->val);
+	} else {
+		return mpz_cmp( x_data->mpz, y_data->mpz );
+	}
 }
 
 int intobj_cmp_long(Object *x, long y)
 {
 	assert(x->type == &intobj_type);
-	return mpz_cmp_si( ((struct IntData*)x->data.val)->mpz, y );
+	struct IntData *data = (struct IntData*) x->data.val;
+
+	if (data->spilled) {
+		return mpz_cmp_si( ((struct IntData*)x->data.val)->mpz, y );
+	} else {
+		return (data->val > y) - (data->val < y);
+	}
+}
+
+static void intobj_spill(Object *obj) {
```

Comment:
Wasn't sure of a better way to do it, anything that spills shouldn't be cached anyways and spilling does not change the value of an int

Reply:
Is this because you want to do a calculation with mpz's? I see a couple ways to do this without mutating immutable things:
- Don't use a union for `val`. A non-spilled integer can have two `vals`, an `mpz_t` for GMP calculations and a `long` for non-GMP calculations.
- Just use `mpz_add_ui`, `mpz_sub_ui` etc when one of the integers is spilled and the other isn't? This creates less unnecessary `mpz`s so it's likely faster anyway.

Also, this is a `static` function so you don't need to prefix it with `intobj_` to avoid name clashes.